### PR TITLE
add basic coredump generation

### PR DESCRIPTION
**Pull request by xtuc**
This change adds a basic coredump generation after a WebAssembly trap was entered. The coredump includes rudimentary stack / process debugging information.

A new CLI argument is added to enable coredump generation:
```
wasmtime --coredump-on-trap=/path/to/coredump/file module.wasm
```

See ./docs/examples-coredump.md for a working example.

Refs https://github.com/bytecodealliance/wasmtime/issues/5732

### CODE REVIEW

File: src/commands/run.rs [hunk 1]
```diff
@@ -170,6 +176,15 @@ impl RunCommand {
 
         let preopen_sockets = self.compute_preopen_sockets()?;
 
+        // Validate coredump-on-trap argument
+        if let Some(coredump_path) = self.coredump_on_trap.as_ref() {
+            if coredump_path.contains("%") {
+                return Err(anyhow!(
+                    "The coredump-on-trap path does not support patterns yet."
```
Comment (xtuc) [inline src/commands/run.rs:183]: For context, unix kernels allows to template the coredump path (https://sigquit.wordpress.com/2009/03/13/the-core-pattern/) using variables starting with %. We might want to support them in the future.

File: Cargo.toml [hunk 2]
```diff
@@ -42,6 +42,7 @@ wat = { workspace = true }
 serde = "1.0.94"
 serde_json = "1.0.26"
 wasmparser = { workspace = true }
+wasm-coredump-builder = { version = "0.1.9" }
```
Comment (alexcrichton) [inline Cargo.toml:45]: This is also visible on CI, but we're stricter-than-average about the dependencies we pull in for Wasmtime. Relying on something external here is fine but this specifically pulls in a number of things which I think we'll want to change/address:

* `colored` - in theory this shouldn't be necessary for just building core dumps
* `core-wasm-ast` - this looks like it may be duplicating the functionality of the already-in-use `wasmparser` crate perhaps? If that's the case we'd prefer to only have one definition of parsing wasm within wasmtime. 
* `hermit-abi` - I think this is due to the `num_cpus` update, but is the update required?
* `threadpool` - I think in theory this shouldn't be necessary for a core dump builder
* `wasm-printer` - does this perhaps duplicate the functionality already present in `wasmprinter`? If so like `wasmparser` it would be best to only have one implementation of this rather than multiple.

I'll also note that we're also already using the `wasm-encoder` crate for creating core wasm modules, so if core dump functionality could be added to that or somehow leverage that I think that would be ideal.

File: src/commands/run.rs [hunk 3]
```diff
@@ -170,6 +176,15 @@ impl RunCommand {
 
         let preopen_sockets = self.compute_preopen_sockets()?;
 
+        // Validate coredump-on-trap argument
+        if let Some(coredump_path) = self.coredump_on_trap.as_ref() {
+            if coredump_path.contains("%") {
+                return Err(anyhow!(
```
Comment (alexcrichton) [inline src/commands/run.rs:182]: You can shorten this a bit with just `bail!("...")` if you'd like

File: src/commands/run.rs [hunk 4]
```diff
@@ -376,13 +391,54 @@ impl RunCommand {
...
             }
-        })?;
+        });
+        match invoke_res {
+            Ok(_) => {}
+            Err(err) => {
+                if let Some(coredump_path) = self.coredump_on_trap.as_ref() {
+                    let bt = err
+                        .downcast_ref::<WasmBacktrace>()
+                        .ok_or_else(|| anyhow!("failed to downcast to WasmBacktrace"))?;
```
Comment (alexcrichton) [inline src/commands/run.rs:407]: This'll want to be a fallible conversion, only emitting a backtrace in the core dump if `WasmBacktrace` is available I think. (as opposed to returning a different kind of error if the backtrace is missing)

File: src/commands/run.rs [hunk 5]
```diff
@@ -376,13 +391,54 @@ impl RunCommand {
...
             if let Some(name) = name {
                 format!("failed to invoke `{}`", name)
             } else {
                 format!("failed to invoke command default")
             }
-        })?;
+        });
+        match invoke_res {
+            Ok(_) => {}
+            Err(err) => {
```
Comment (alexcrichton) [inline src/commands/run.rs:403]: I think this logic below may want to be conditional on a `err.is::<wasmtime::Trap>()` condition since there's a number of other unrelated things that could go wrong during execution.

File: src/commands/run.rs [hunk 6]
```diff
@@ -376,13 +391,54 @@ impl RunCommand {
...
+                    }
+
+                    let coredump = coredump_builder
+                        .serialize()
+                        .map_err(|err| anyhow!("failed to serialize coredump: {}", err))?;
+
+                    let mut f = File::create(coredump_path)
+                        .context(format!("failed to create file at `{}`", coredump_path))?;
+                    f.write_all(&coredump)
+                        .context("failed to write coredump file")?;
```
Comment (alexcrichton) [inline src/commands/run.rs:433]: Handling errors here is a bit interesting because while these operations are fallible I don't think that the error here wants to necessarily "overwrite" the trap error. I also don't know of a great way to "fuse" errors either. Perhaps for tnow this could log a warning message or similar?

File: src/commands/run.rs [hunk 7]
```diff
@@ -376,13 +391,54 @@ impl RunCommand {
...
+                        .executable_name(&self.module.to_str().unwrap_or_else(|| "unknown"));
+
+                    {
+                        let mut thread_builder =
+                            wasm_coredump_builder::ThreadBuilder::new().thread_name("main");
+
+                        for frame in bt.frames() {
+                            let coredump_frame = wasm_coredump_builder::FrameBuilder::new()
+                                .funcidx(frame.func_index())
+                                .build();
```
Comment (alexcrichton) [inline src/commands/run.rs:419]: The backtrace frames here should also have offset information to get to the precise opcode if you'd like to add that.

**Review by alexcrichton — COMMENTED** [review]
Thanks!

One thought I have with this is that while placing it in the CLI is perhaps a good starting point we'll likely want this in a more libararified form at some point too because embeddings of Wasmtime don't necessarily all use the CLI. This current integration is also somewhat limited because the native call stack is gone and as a result impossible to ever recover locals and their information. Basically the "ideal" location for a core dump is probably at the place of the trap itself, which means that this wouldn't be possible to do from the CLI but rather would be required to be baked-in as well. Additionally being baked-in I think will be required to get access to non-exported globals/memories/etc.

None of that is to say though that this shouldn't be added here at this time. Moreso that a more "final" integration of core dumps I think will probably look significantly differently architected which isn't a problem but perhaps worth considering.

File: Cargo.toml (hunk 2 above)
Comment (xtuc) [reply Cargo.toml:45]: `colored`, `threadpool` and `num_cpus`  are dependencies not directly used by `wasm-coredump-builder`, I'll make sure to remove them.

`core-wasm-ast` (in theory, just defines the stucts to represent coredump) and `wasm-printer` (maybe a bad name, it's used to serialize the coredump to binary Wasm) are required by `wasm-coredump-builder`. 

> I'll also note that we're also already using the wasm-encoder crate for creating core wasm modules, so if core dump functionality could be added to that or somehow leverage that I think that would be ideal.

That's indeed duplicated features with my crates. I believe I can rewrite `wasm-coredump-builder` to only use `wasm-encoder`. A disadvantage of that approach on my side is that it won't integrate well with my tools / crates, but could be specific to wasmtime.

File: src/commands/run.rs (hunk 4 above)
Comment (xtuc) [reply src/commands/run.rs:407]: Good point. Same with the errors bellow, I'll change the code to allow coredump generation to fail gracefully.

File: src/commands/run.rs (hunk 7 above)
Comment (xtuc) [reply src/commands/run.rs:419]: Yeah, we'll probably want that in the future. I haven't changed (or decided) the coredump format to allow `funcidx` in addition to `codeoffset` yet.

File: Cargo.toml (hunk 2 above)
Comment (xtuc) [reply Cargo.toml:45]: With the libararified / modular CLI approach, dependecies wouldn't matter as much right?

Comment (alexcrichton) [reply Cargo.toml:45]: To clarify I don't mean to pass judgement on which one should be used per se, but within the context of Wasmtime's dependency graph we'd prefer to keep only one implementation of things rather than bringing together various islands of ecosystems. I don't mean to say that you should rewrite crates on top of the crates in the `wasm-tools` repository (although feedback on those is always appreciated), but I believe that `wasm-encoder` is high-level enough and the core dump format is simple enough that doing it "inline" as opposed to using an external crate wouldn't be the end of the world (not that using an external crate is wrong, again just trying to keep the dependencies in Wasmtime under control)

Comment (xtuc) [reply Cargo.toml:45]: yeah I understand, I wasn't trying to argue against. I'm planning to rewrite `wasm-coredump-builder` to depend only on a new coredump specific encoding crate + `wasm-encoder`. That should address the concern.

Comment (xtuc) [reply Cargo.toml]: I published a new version of `wasm-coredump-builder` that depends on `wasm-encoder` and two coredump specific crates:
- https://github.com/xtuc/wasm-coredump/tree/main/lib/coredump-encoder
- https://github.com/xtuc/wasm-coredump/tree/main/lib/coredump-types

I'm not sure why `num_cpus` has been updated. Is it still related to my change?

**Comment by xtuc** [comment]
I changed the coredump generation to gracefully handle errors, as you pointed out. The dependencies are now trimmed down.

File: supply-chain/audits.toml [hunk 8]
```diff
@@ -762,6 +762,21 @@ is similar to what it once was back then. Skimming over the crate there is
...
+[[audits.wasm-coredump-encoder]]
+who = "Sven Sauleau <sven@cloudflare.com>"
+criteria = "safe-to-deploy"
+version = "0.1.10"
+
+[[audits.wasm-coredump-types]]
+who = "Sven Sauleau <sven@cloudflare.com>"
+criteria = "safe-to-deploy"
+version = "0.1.10"
+
```
Comment (alexcrichton) [inline supply-chain/audits.toml:779]: For now we're enforcing that the edits to the `supply-chain/*` folder are only done by "trusted reviewers" of the project since this affects our supply-chain security (and others who import these audits). To that end thanks for adding these here, but our current process (which isn't well documented) is that if a contributor needs new vet entries we [post a separate PR](https://github.com/bytecodealliance/wasmtime/pull/5878) which the contributor's PR is then rebased on top of.

Once that PR lands, though, you should be able to rebase this PR on top of that one and back out the changes to the `supply-chain/*` directory.

File: src/commands/run.rs [hunk 9]
```diff
@@ -557,3 +591,36 @@ fn ctx_set_listenfd(num_fd: usize, builder: WasiCtxBuilder) -> Result<(usize, Wa
...
+    let mut coredump_builder =
+        wasm_coredump_builder::CoredumpBuilder::new().executable_name(source_name);
+
+    {
+        let mut thread_builder = wasm_coredump_builder::ThreadBuilder::new().thread_name("main");
+
+        for frame in bt.frames() {
+            let coredump_frame = wasm_coredump_builder::FrameBuilder::new()
+                .funcidx(frame.func_index())
+                .build();
```
Comment (alexcrichton) [inline src/commands/run.rs:609]: Sorry the context is lost on the prior comment with the update here, but again while this isn't a blocker not including the function offset seems like it could hinder the usefulness of backtraces since you don't get where you are in a function, just that you're in a certain function. If you're exploring a certian `unreachable` condition, for example, there'd be no way to figure out which one in a function was hit.

File: src/commands/run.rs [hunk 10]
```diff
@@ -557,3 +591,36 @@ fn ctx_set_listenfd(num_fd: usize, builder: WasiCtxBuilder) -> Result<(usize, Wa
 
     Ok((num_fd, builder))
 }
+
+fn generate_coredump(err: &anyhow::Error, source_name: &str, coredump_path: &str) -> Result<()> {
+    let bt = err
+        .downcast_ref::<wasmtime::WasmBacktrace>()
+        .ok_or_else(|| anyhow!("failed to downcast to WasmBacktrace"))?;
```
Comment (alexcrichton) [inline src/commands/run.rs:598]: Since this will be a user-facing error message I think it might be better to reword this as "no wasm backtrace found to generate core dump with" or something along those lines.

File: src/commands/run.rs [hunk 11]
```diff
@@ -557,3 +591,36 @@ fn ctx_set_listenfd(num_fd: usize, builder: WasiCtxBuilder) -> Result<(usize, Wa
...
+                .build();
+            thread_builder.add_frame(coredump_frame);
+        }
+
+        coredump_builder.add_thread(thread_builder.build());
+    }
+
+    let coredump = coredump_builder
+        .serialize()
+        .map_err(|err| anyhow!("failed to serialize coredump: {}", err))?;
```
Comment (alexcrichton) [inline src/commands/run.rs:618]: Could the `.map_err` here become `.context(..)`?

File: src/commands/run.rs [hunk 12]
```diff
@@ -557,3 +591,36 @@ fn ctx_set_listenfd(num_fd: usize, builder: WasiCtxBuilder) -> Result<(usize, Wa
...
+
+    let mut f = File::create(coredump_path)
+        .context(format!("failed to create file at `{}`", coredump_path))?;
+    f.write_all(&coredump)
+        .context("failed to write coredump file")?;
```
Comment (alexcrichton) [inline src/commands/run.rs:623]: Could this error also contain the path name?

File: src/commands/run.rs [hunk 13]
```diff
@@ -170,6 +176,13 @@ impl RunCommand {
 
         let preopen_sockets = self.compute_preopen_sockets()?;
 
+        // Validate coredump-on-trap argument
+        if let Some(coredump_path) = self.coredump_on_trap.as_ref() {
+            if coredump_path.contains("%") {
+                bail!("The coredump-on-trap path does not support patterns yet.")
```
Comment (alexcrichton) [inline src/commands/run.rs:182]: As a minor nit stylistically our error messages start with lowercase letters (including the "Core dumped at ..." message below)

File: src/commands/run.rs [hunk 14]
```diff
@@ -376,13 +389,34 @@ impl RunCommand {
...
+        });
+
+        if let Err(err) = invoke_res {
+            let err = if err.is::<wasmtime::Trap>() {
+                if let Some(coredump_path) = self.coredump_on_trap.as_ref() {
+                    let source_name = self.module.to_str().unwrap_or_else(|| "unknown");
+
+                    if let Err(err) = generate_coredump(&err, &source_name, coredump_path) {
+                        eprintln!("warning: coredump failed to generate: {}", err);
+                        err
```
[the remark this answers is not in the document]
Comment (xtuc) [reply src/commands/run.rs:407]: Yes, good catch!

File: supply-chain/audits.toml (hunk 8 above)
Comment (xtuc) [reply supply-chain/audits.toml:779]: Ok, thanks.

File: src/commands/run.rs (hunk 11 above)
Comment (xtuc) [reply src/commands/run.rs:618]: The error is from `wasm-coredump-builder` and is type `Box<dyn std::error::Error + Sync + Send>` instead of the usual `anyhow::Error`. Is that a problem?

File: src/commands/run.rs (hunk 9 above)
Comment (xtuc) [reply src/commands/run.rs:610]: I opened https://github.com/WebAssembly/tool-conventions/pull/202 to add `funcidx` to the coredump format.

Comment (xtuc) [reply src/commands/run.rs:610]: unfortunately, I don't see a way to retrieve the instruction offset in https://docs.rs/wasmtime/6.0.0/wasmtime/struct.FrameInfo.html.

There's a  method to retrieve the func offset in the module: https://docs.rs/wasmtime/6.0.0/wasmtime/struct.FrameInfo.html#method.func_offset.

Comment (cfallin) [reply src/commands/run.rs:610]: It looks like `func_offset` is actually what is needed? The documentation reads "Returns the offset from the original wasm module’s function to this frame’s program counter." i.e., it encodes `PC - function_start`, the offset of the instruction. (The name is somewhat confusing though -- "offset in function" rather than "offset of function"!)

Comment (xtuc) [reply src/commands/run.rs:610]: Ah great, I misunderstood. Thanks

Comment (xtuc) [reply src/commands/run.rs]: Added `codeoffset`.

File: src/commands/run.rs [hunk 15]
```diff
@@ -557,3 +592,37 @@ fn ctx_set_listenfd(num_fd: usize, builder: WasiCtxBuilder) -> Result<(usize, Wa
...
+
+    let mut coredump_builder =
+        wasm_coredump_builder::CoredumpBuilder::new().executable_name(source_name);
+
+    {
+        let mut thread_builder = wasm_coredump_builder::ThreadBuilder::new().thread_name("main");
+
+        for frame in bt.frames() {
+            let coredump_frame = wasm_coredump_builder::FrameBuilder::new()
+                .codeoffset(frame.func_offset().unwrap_or(0) as u32)
```
Comment (alexcrichton) [inline src/commands/run.rs:609]: Should this perhaps be a conditional invocation of `.codeoffset` due to it not always being available?

File: src/commands/run.rs (hunk 11 above)
Comment (alexcrichton) [reply src/commands/run.rs:620]: It's a problem insofar as the error's contexet is all lost here, only displaying one error in a possible chain of errors contained within the trait object. Whether your library exercises that I'm not sure, though.

**Review by alexcrichton — APPROVED** [review]
I've added the new versions to https://github.com/bytecodealliance/wasmtime/pull/5894

File: src/commands/run.rs (hunk 11 above)
Comment (xtuc) [reply src/commands/run.rs:620]: It's not a problem for my library, for now, as it doesn't add context to errors.

File: src/commands/run.rs (hunk 15 above)
Comment (xtuc) [reply src/commands/run.rs:609]: The result would be the same; `codeoffset` has to be specified in coredump and can default to 0 if unknown. At least now, it's explicit.
